mlflow/types/type_hints.py

```diff
@@ -0,0 +1,297 @@
+                )
+            )
+    if invalid_fields:
+        raise MlflowException.invalid_parameter_value(
+            "The following fields in the Pydantic model do not have type annotations: "
+            f"{invalid_fields}. Please add type annotations to these fields."
+        )
+
+    return output_fields
+
+
+def _is_pydantic_type_hint(type_hint: type[Any]) -> bool:
+    try:
+        return inspect.isclass(type_hint) and issubclass(type_hint, pydantic.BaseModel)
+    # inspect.isclass(dict[str, int]) is True, but issubclass raises a TypeError
+    except TypeError:
+        return False
+
+
+def model_fields(model: pydantic.BaseModel) -> dict[str, pydantic.fields.FieldInfo]:
+    if PYDANTIC_V2:
+        return model.model_fields
+    return model.__fields__
+
+
+def model_validate(model: pydantic.BaseModel, values: Any) -> None:
+    if PYDANTIC_V2:
+        # use strict mode to avoid any data conversion here
+        # e.g. "123" will not be converted to 123 if the type is int
+        model.model_validate(values, strict=True)
+    else:
+        model.validate(values)
+
+
+def _infer_schema_from_type_hint(type_hint: type[Any]) -> Schema:
+    if _is_pydantic_type_hint(type_hint):
+        col_specs = _infer_fields_from_pydantic_model(type_hint, "ColSpec")
+        return Schema(inputs=col_specs)
+    else:
+        col_spec_type = _infer_colspec_type_from_type_hint(type_hint)
+        # Creating Schema with unnamed optional inputs is not supported
+        if col_spec_type.required is False:
+            raise MlflowException.invalid_parameter_value(
+                "Optional type hint is not supported, it can only be used as a field type in "
+                "Pydantic models."
+            )
+        return Schema([ColSpec(type=col_spec_type.dtype, required=col_spec_type.required)])
+
+
+def _validate_example_against_type_hint(example: Any, type_hint: type[Any]) -> None:
+    if _is_pydantic_type_hint(type_hint):
+        try:
+            model_validate(type_hint, example)
+        except pydantic.ValidationError as e:
+            raise MlflowException.invalid_parameter_value(
+                message=f"Input example is not valid for Pydantic model `{type_hint.__name__}`",
+            ) from e
+    elif type_hint == Any:
+        return
+    elif type_hint in TYPE_HINTS_TO_DATATYPE_MAPPING:
+        if isinstance(example, type_hint):
+            return
+        raise MlflowException.invalid_parameter_value(
+            f"Expected type {type_hint}, but got {type(example).__name__}"
+        )
+    elif origin_type := get_origin(type_hint):
+        args = get_args(type_hint)
+        if origin_type is list:
+            return _validate_list_elements(element_type=args[0], example=example)
+        elif origin_type is dict:
+            return _validate_dict_elements(element_type=args[1], example=example)
+        elif origin_type == Union:
+            # Optional type
+            if type(None) in args:
```

Comment:
```suggestion
NONE_TYPE = type(None)
...
            if NONE_TYPE in args:
```

Can we define `NONE_TYPE`?